pkg/client/didexchange/client.go

```diff
@@ -119,35 +119,144 @@ func (c *Client) RemoveConnection(id string) error {
-func (c *Client) startServiceEventListener() error {
-	// register the action event channel
+func (c *Client) startServiceEventListener() {
+	// listen for action event and message events
+	go func() {
+		for {
+			select {
+			case e := <-c.actionCh:
+				// assigned to var as lint fails with : Using a reference for the variable on range scope (scopelint)
+				msg := e
+				c.handleActionEvent(&msg)
+			case e := <-c.msgCh:
+				// assigned to var as lint fails with : Using a reference for the variable on range scope (scopelint)
+				msg := e
+				c.handleMessageEvent(&msg)
+			}
+		}
+	}()
+}
+
+// RegisterActionEvent on DID Exchange protocol messages. The events are triggered for incoming exchangeRequest,
+// exchangeResponse and exchangeAck message types. The consumer need to invoke the callback to resume processing.
+// Only one channel can be registered for the action events. The function will throw error if a channel is already
+// registered. The AutoExecuteActionEvent() function can be used to automatically trigger callback function for the
+// event.
+func (c *Client) RegisterActionEvent(ch chan<- dispatcher.DIDCommAction) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.actionEvent != nil {
+		return errors.New("channel is already registered for the action event")
+	}
+
+	// register the service action event channel
 	err := c.didexchangeSvc.RegisterActionEvent(c.actionCh)
 	if err != nil {
 		return fmt.Errorf("didexchange action event registration failed: %w", err)
 	}
 
-	// register the message event channel
-	err = c.didexchangeSvc.RegisterMsgEvent(c.msgCh)
+	c.actionEvent = ch
+
+	return nil
+}
+
+// UnregisterActionEvent on DID Exchange protocol messages. Refer RegisterActionEvent().
+func (c *Client) UnregisterActionEvent(ch chan<- dispatcher.DIDCommAction) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.actionEvent != ch {
+		return errors.New("invalid channel passed to unregister the action event")
+	}
+
+	// unregister the service action event channel
+	err := c.didexchangeSvc.UnregisterActionEvent(c.actionCh)
 	if err != nil {
-		return fmt.Errorf("didexchange message event registration failed: %w", err)
+		return fmt.Errorf("didexchange action event unregistration failed: %w", err)
 	}
 
-	// TODO https://github.com/hyperledger/aries-framework-go/issues/199 - Generate Client events
-	// for now, auto execute the actions
-	go func() {
-		err := didexchange.AutoExecuteActionEvent(c.actionCh)
+	c.actionEvent = nil
+
+	return nil
+}
+
+// RegisterMsgEvent on DID Exchange protocol messages. The message events are triggered for state transitions. Client
+// will not expect any callback on these events unlike Action events.
+func (c *Client) RegisterMsgEvent(ch chan<- dispatcher.StateMsg) error {
+	c.msgEventLock.Lock()
+	count := len(c.msgEvents)
+	c.msgEventLock.Unlock()
+
+	// register the service message event channel (only once)
+	if count == 0 {
```

Comment:
I just realized that there could be a race condition here. If two concurrent calls come here then both of them could get a count of 0 and then both would register the event.

Reply:
The original code had creation at startup. Reverted this logic.